File: astro-app/src/lib/extractor/html-extractor.ts
import * as cheerio from 'cheerio';
import type { ScraperMapping, FieldMapping } from './mapping-loader.js';
import { findByName } from './mapping-loader.js';
import { retrieveTargetText } from './strategies.js';
import { extractImages } from './image-extractor.js';
import { extractFeatures } from './feature-extractor.js';
import { booleanEvaluators } from './field-processors.js';
import {
  assessQualityWeighted,
  getFieldImportance,
  type QualityGrade,
  type QualityAssessment,
  type FieldResult,
} from './quality-scorer.js';
import { splitPropertyHash, type SplitSchema } from './schema-splitter.js';
import { computeFingerprint } from '../services/listing-fingerprint.js';
import { sanitizePropertyHash } from '../services/content-sanitizer.js';

export interface FieldTrace {
  field: string;
  section: string;
  strategy: string;
  rawText: string;
  value: unknown;
  fallbackUsed?: number;
}

export interface ContentAnalysis {
  htmlLength: number;
  hasScriptTags: boolean;
  jsonLdCount: number;
  scriptJsonVarsFound: string[];
  appearsBlocked: boolean;
  appearsJsOnly: boolean;
}

export interface ExtractionDiagnostics {
  scraperName: string;
  fieldTraces: FieldTrace[];
  totalFields: number;
  populatedFields: number;
  emptyFields: string[];
  extractableFields: number;
  populatedExtractableFields: number;
  extractionRate: number;
  qualityGrade: QualityGrade;
  qualityLabel: string;
  successClassification: 'excellent' | 'good' | 'partial' | 'failed';
  expectedExtractionRate?: number;
  expectedQualityGrade?: QualityGrade;
  meetsExpectation: boolean;
  expectationGap?: number;
  expectationStatus?: 'unknown' | 'above' | 'meets' | 'below' | 'well_below';
  weightedExtractionRate?: number;
  criticalFieldsMissing?: string[];
  contentAnalysis?: ContentAnalysis;
}

export type ExtractionStatus = 'full' | 'partial' | 'blocked' | 'failed';

export interface ExtractionResult {
  success: boolean;
  status: ExtractionStatus;
  properties: Record<string, unknown>[];
  errorMessage?: string;
  warnings: string[];
  diagnostics?: ExtractionDiagnostics;
  splitSchema?: SplitSchema;
  fingerprint?: string;
}

export interface ExtractParams {
  html: string;
  sourceUrl: string;
  scraperMappingName?: string;
  scraperMapping?: ScraperMapping;
}

/**
 * Resolves the scraper mapping to use.
 * Port of Ruby HtmlExtractor#resolve_mapping.
 */
function resolveMapping(
  mapping?: ScraperMapping,
  name?: string,
  _host?: string
): ScraperMapping {
  if (mapping) return mapping;
  if (name) {
    const found = findByName(name);
    if (!found) throw new Error(`Unknown scraper mapping: ${name}`);
    return found;
  }
  throw new Error('No scraper mapping provided and auto-detection not implemented in PoC');
}

function describeStrategy(m: FieldMapping): string {
  if (m.flightDataPath) return `flightDataPath:${m.flightDataPath}`;
  if (m.jsonLdPath) return `jsonLdPath:${m.jsonLdPath}`;
  if (m.scriptJsonPath && m.scriptJsonVar) return `scriptJsonPath:${m.scriptJsonVar}.${m.scriptJsonPath}`;
  if (m.scriptRegEx) return `scriptRegEx:${m.scriptRegEx.slice(0, 40)}`;
  if (m.urlPathPart) return `urlPathPart:${m.urlPathPart}`;
  if (m.cssLocator) return `cssLocator:${m.cssLocator}`;
  if (m.value) return `default:${m.value}`;
  return 'none';
}

const KNOWN_SCRIPT_VARS = ['PAGE_MODEL', '__NEXT_DATA__', '__INITIAL_STATE__', 'dataLayer'];

/**
 * Analyze content for provenance tracking.
 */
function analyzeContent($: cheerio.CheerioAPI, html: string): ContentAnalysis {
  const scriptTags = $('script');
  const hasScriptTags = scriptTags.length > 0;
  const jsonLdCount = $('script[type="application/ld+json"]').length;

  const scriptJsonVarsFound: string[] = [];
  let scriptText: string | null = null;
  const getScriptText = (): string => {
    if (scriptText === null) {
      scriptText = scriptTags.text();
    }
    return scriptText;
  };
  for (const varName of KNOWN_SCRIPT_VARS) {
    const pattern = new RegExp(`(?:window\\.)?${varName}\\s*=`);
    if (pattern.test(getScriptText()) || $(`script#${varName}`).length > 0) {
      scriptJsonVarsFound.push(varName);
    }
  }

  const bodyText = $('body').text().trim();
  const bodyLength = bodyText.length;
  const divCount = $('div').length;

  // Bot block detection: short body + captcha/verification keywords
  const blockKeywords = /captcha|verify|access denied|cloudflare|just a moment/i;
  const appearsBlocked = bodyLength < 500 && blockKeywords.test(bodyText);

  // JS-only shell detection: short body text + many scripts + few divs
  const appearsJsOnly = bodyLength < 200 && scriptTags.length > 5 && divCount < 10;

  return {
    htmlLength: html.length,
    hasScriptTags,
    jsonLdCount,
    scriptJsonVarsFound,
    appearsBlocked,
    appearsJsOnly,
  };
}

/**
 * Main entry point: extract structured property data from raw HTML.
 * Port of Ruby HtmlExtractor.call.
 *
 * Pure function with zero network I/O.
 */
export function extractFromHtml(params: ExtractParams): ExtractionResult {
  const { html, sourceUrl, scraperMappingName, scraperMapping: providedMapping } = params;

  let mapping: ScraperMapping;
  try {
    mapping = resolveMapping(providedMapping, scraperMappingName);
  } catch (err: unknown) {
    return {
      success: false,
      status: 'failed' as ExtractionStatus,
      properties: [],
      warnings: [],
      errorMessage: err instanceof Error ? err.message : String(err),
    };
  }

  const warnings: string[] = [];

  const $ = cheerio.load(html);
  const uri = new URL(sourceUrl);
  const propertyHash: Record<string, unknown> = {};
  const traces: FieldTrace[] = [];

  // Content analysis (provenance tracking)
  const contentAnalysis = analyzeContent($, html);

  // Default values
  if (mapping.defaultValues) {
    for (const [key, fieldMapping] of Object.entries(mapping.defaultValues)) {
      propertyHash[key] = fieldMapping.value;
      traces.push({
        field: key,
        section: 'defaultValues',
        strategy: describeStrategy(fieldMapping),
        rawText: fieldMapping.value || '',
        value: propertyHash[key],
      });
    }
  }

  // Images
  if (mapping.images) {
    for (const imageMapping of mapping.images) {
      propertyHash['image_urls'] = extractImages($, html, imageMapping, uri);
      const images = propertyHash['image_urls'] as Array<{ url: string }>;
      traces.push({
        field: 'image_urls',
        section: 'images',
        strategy: describeStrategy(imageMapping),
        rawText: images.length > 0 ? `${images.length} images` : '',
        value: images,
      });
    }
  }

  // Features
  if (mapping.features) {
    for (const featureMapping of mapping.features) {
      propertyHash['features'] = extractFeatures($, featureMapping, uri);
      const feats = propertyHash['features'] as string[];
      traces.push({
        field: 'features',
        section: 'features',
        strategy: describeStrategy(featureMapping),
        rawText: feats.length > 0 ? `${feats.length} features` : '',
        value: feats,
      });
    }
  }

  // Int fields
  if (mapping.intFields) {
    for (const [key, fieldMapping] of Object.entries(mapping.intFields)) {
      const result = retrieveTargetText($, html, fieldMapping, uri);
      propertyHash[key] = parseInt(result.text.trim(), 10) || 0;
      traces.push({
        field: key,
        section: 'intFields',
        strategy: result.strategyDescription,
        rawText: result.text.slice(0, 120),
        value: propertyHash[key],
        fallbackUsed: result.strategyIndex > 0 ? result.strategyIndex : undefined,
      });
    }
  }

  // Float fields
  if (mapping.floatFields) {
    for (const [key, fieldMapping] of Object.entries(mapping.floatFields)) {
      const result = retrieveTargetText($, html, fieldMapping, uri);
      let text = result.text;
      if (fieldMapping.stripPunct) {
        text = text.replace(/\./g, '').replace(/,/g, '');
      }
      if (fieldMapping.stripFirstChar) {
        text = text.trim().slice(1) || '';
      }
      propertyHash[key] = parseFloat(text.trim()) || 0;
      traces.push({
        field: key,
        section: 'floatFields',
        strategy: result.strategyDescription,
        rawText: text.slice(0, 120),
        value: propertyHash[key],
        fallbackUsed: result.strategyIndex > 0 ? result.strategyIndex : undefined,
      });
    }
  }

  // Text fields
  if (mapping.textFields) {
    for (const [key, fieldMapping] of Object.entries(mapping.textFields)) {
      const result = retrieveTargetText($, html, fieldMapping, uri);
      propertyHash[key] = result.text.trim();
      traces.push({
        field: key,
        section: 'textFields',
        strategy: result.strategyDescription,
        rawText: result.text.slice(0, 120),
        value: propertyHash[key],
        fallbackUsed: result.strategyIndex > 0 ? result.strategyIndex : undefined,
      });
    }
  }

  // Boolean fields
  if (mapping.booleanFields) {
    for (const [key, fieldMapping] of Object.entries(mapping.booleanFields)) {
      const result = retrieveTargetText($, html, fieldMapping, uri);
      let text = result.text;
      let evaluatorParam = fieldMapping.evaluatorParam || '';

      if (fieldMapping.caseInsensitive) {
        text = text.toLowerCase();
        evaluatorParam = evaluatorParam.toLowerCase();
      }

      const evaluatorFn = fieldMapping.evaluator
        ? booleanEvaluators[fieldMapping.evaluator]
        : undefined;

      propertyHash[key] = evaluatorFn
        ? evaluatorFn(text.trim(), evaluatorParam)
        : false;
      traces.push({
        field: key,
        section: 'booleanFields',
        strategy: result.strategyDescription,
        rawText: text.slice(0, 120),
        value: propertyHash[key],
        fallbackUsed: result.strategyIndex > 0 ? result.strategyIndex : undefined,
      });
    }
  }

  const emptyFields = traces
    .filter(t => t.rawText === '' || t.value === 0 || t.value === false || t.value === '')
    .map(t => t.field);

  // Quality scoring: exclude defaultValues to measure actual extraction success
  const extractableTraces = traces.filter(t => t.section !== 'defaultValues');
  const populatedExtractable = extractableTraces.filter(
    t => t.rawText !== '' && t.value !== 0 && t.value !== false && t.value !== ''
  );
  const extractableFields = extractableTraces.length;
  const populatedExtractableFields = populatedExtractable.length;
  const extractionRate = extractableFields > 0 ? populatedExtractableFields / extractableFields : 0;

  // Build FieldResult array for weighted scoring
  const fieldResults: FieldResult[] = extractableTraces.map(t => ({
    field: t.field,
    populated: t.rawText !== '' && t.value !== 0 && t.value !== false && t.value !== '',
    importance: getFieldImportance(t.field),
  }));

  const quality: QualityAssessment = assessQualityWeighted(fieldResults, mapping.expectedExtractionRate);

  const successClassification: ExtractionDiagnostics['successClassification'] =
    quality.grade === 'A' ? 'excellent'
      : quality.grade === 'B' ? 'good'
        : quality.grade === 'C' ? 'partial'
          : 'failed';

  // Sanitize extracted data (strip HTML, validate URLs)
  const sanitizedHash = sanitizePropertyHash(propertyHash);

  // Split schema
  const split = splitPropertyHash(sanitizedHash);

  const diagnostics: ExtractionDiagnostics = {
    scraperName: mapping.name,
    fieldTraces: traces,
    totalFields: traces.length,
    populatedFields: traces.length - emptyFields.length,
    emptyFields,
    extractableFields,
    populatedExtractableFields,
    extractionRate,
    qualityGrade: quality.grade,
    qualityLabel: quality.label,
    successClassification,
    expectedExtractionRate: mapping.expectedExtractionRate,
    expectedQualityGrade: quality.expectedGrade,
    meetsExpectation: quality.meetsExpectation,
    expectationGap: quality.expectationGap,
    expectationStatus: quality.expectationStatus,
    weightedExtractionRate: quality.weightedRate,
    criticalFieldsMissing: quality.criticalFieldsMissing,
    contentAnalysis,
  };

  // Collect warnings
  if (contentAnalysis.appearsBlocked) {
    warnings.push('Page appears to be blocked by bot detection');
  }
  if (contentAnalysis.appearsJsOnly) {
    warnings.push('Page appears to be a JS-only shell — content may require browser rendering');
  }
  if (quality.criticalFieldsMissing && quality.criticalFieldsMissing.length > 0) {
    warnings.push(`Critical fields missing: ${quality.criticalFieldsMissing.join(', ')}`);
  }

  // Determine extraction status
  let extractionStatus: ExtractionStatus;
  if (contentAnalysis.appearsBlocked) {
    extractionStatus = 'blocked';
  } else if (extractableFields > 0 && populatedExtractableFields === 0) {
    extractionStatus = 'failed';
  } else if (quality.criticalFieldsMissing && quality.criticalFieldsMissing.length > 0) {
    extractionStatus = 'partial';
  } else if (quality.grade === 'A' || quality.grade === 'B') {
    extractionStatus = 'full';
  } else {
    extractionStatus = 'partial';
  }

  // Compute content fingerprint for deduplication
  const fingerprint = computeFingerprint({
    title: typeof propertyHash.title === 'string' ? propertyHash.title : undefined,
    price_float: typeof propertyHash.price_float === 'number' ? propertyHash.price_float : undefined,
    address_string: typeof propertyHash.address_string === 'string' ? propertyHash.address_string : undefined,
  });

  console.log(`[Extractor] ${mapping.name}: Grade ${quality.grade} (${quality.label}) — ${populatedExtractableFields}/${extractableFields} extractable fields (${Math.round(extractionRate * 100)}%), weighted ${Math.round((quality.weightedRate || 0) * 100)}% [${extractionStatus}]`);
  if (quality.criticalFieldsMissing && quality.criticalFieldsMissing.length > 0) {
    console.log(`[Extractor] Critical fields missing: ${quality.criticalFieldsMissing.join(', ')}`);
  }
  if (emptyFields.length > 0) {
    console.log(`[Extractor] Empty fields: ${emptyFields.join(', ')}`);
  }
  for (const t of traces) {
    const status = t.rawText ? '\u2713' : '\u2717';
    const fb = t.fallbackUsed != null ? ` [fallback ${t.fallbackUsed}]` : '';
    console.log(`[Extractor]   ${status} ${t.field} (${t.strategy}${fb}) \u2192 ${JSON.stringify(t.value)}`);
  }

  return {
    success: true,
    status: extractionStatus,
    properties: [sanitizedHash],
    warnings,
    diagnostics,
    splitSchema: split,
    fingerprint,
  };
}

File: app/services/property_web_scraper/html_extractor.rb
require 'nokogiri'

module PropertyWebScraper
  # Extracts structured property data from raw HTML using configurable mappings.
  #
  # This is a pure-function service with zero network I/O. It takes
  # fully-rendered HTML (from any source: Chrome extension, headless browser,
  # direct fetch) and returns structured property data.
  #
  # @example Extract from pre-rendered HTML
  #   result = HtmlExtractor.call(
  #     html: "<html>...</html>",
  #     source_url: "https://www.idealista.com/inmueble/123/",
  #     scraper_mapping_name: "idealista"
  #   )
  #   result[:properties].first["title"] #=> "Piso en venta..."
  #
  # @example Extract with a pre-loaded mapping
  #   mapping = ScraperMapping.find_by_name("rightmove")
  #   result = HtmlExtractor.call(
  #     html: page_html,
  #     source_url: url,
  #     scraper_mapping: mapping
  #   )
  class HtmlExtractor
    attr_reader :doc, :uri, :scraper_mapping

    # Main entry point.
    #
    # @param html [String] fully-rendered HTML
    # @param source_url [String] the original page URL (for relative URL resolution + URL-path extraction)
    # @param scraper_mapping_name [String, nil] override mapping name (default: auto-detect from source_url host)
    # @param scraper_mapping [ScraperMapping, nil] pre-loaded mapping object (takes precedence)
    # @return [Hash] { success: Boolean, properties: [Hash], error_message: String? }
    def self.call(html:, source_url:, scraper_mapping_name: nil, scraper_mapping: nil)
      new(html, source_url, scraper_mapping_name, scraper_mapping).extract
    end

    def initialize(html, source_url, scraper_mapping_name, scraper_mapping)
      @uri = URI.parse(source_url)
      @doc = Nokogiri::HTML(html)
      @scraper_mapping = resolve_mapping(scraper_mapping, scraper_mapping_name)
    end

    # Extracts property data from the parsed HTML document.
    #
    # @return [Hash] { success: true, properties: [Hash] }
    def extract
      property_hash = {}

      if scraper_mapping.defaultValues
        scraper_mapping.defaultValues.keys.each do |mapping_key|
          mapping = scraper_mapping.defaultValues[mapping_key]
          property_hash[mapping_key] = mapping['value']
        end
      end

      if scraper_mapping.images
        scraper_mapping.images.each do |image_mapping|
          retrieved_array = retrieve_images_array doc, image_mapping, uri
          property_hash["image_urls"] = retrieved_array
        end
      end

      if scraper_mapping.features
        scraper_mapping.features.each do |feature_mapping|
          retrieved_array = retrieve_features_array doc, feature_mapping, uri
          property_hash["features"] = retrieved_array
        end
      end

      if scraper_mapping.intFields
        scraper_mapping.intFields.keys.each do |mapping_key|
          mapping = scraper_mapping.intFields[mapping_key]
          retrieved_text = retrieve_target_text doc, mapping, uri
          property_hash[mapping_key] = retrieved_text.strip.to_i
        end
      end

      if scraper_mapping.floatFields
        scraper_mapping.floatFields.keys.each do |mapping_key|
          mapping = scraper_mapping.floatFields[mapping_key]
          retrieved_text = retrieve_target_text doc, mapping, uri
          if mapping['stripPunct']
            retrieved_text = retrieved_text.tr('.', '').tr(',', '')
          end
          if mapping['stripFirstChar']
            retrieved_text = retrieved_text.strip[1..] || ''
          end
          property_hash[mapping_key] = retrieved_text.strip.to_f
        end
      end

      if scraper_mapping.textFields
        scraper_mapping.textFields.keys.each do |mapping_key|
          mapping = scraper_mapping.textFields[mapping_key]
          retrieved_text = retrieve_target_text doc, mapping, uri
          property_hash[mapping_key] = retrieved_text.strip
        end
      end

      if scraper_mapping.booleanFields
        scraper_mapping.booleanFields.keys.each do |mapping_key|
          mapping = scraper_mapping.booleanFields[mapping_key]
          retrieved_text = retrieve_target_text doc, mapping, uri
          evaluator_text = mapping['evaluatorParam']

          if mapping['caseInsensitive']
            retrieved_text = retrieved_text.downcase
            evaluator_text = evaluator_text.downcase
          end

          property_hash[mapping_key] = retrieved_text.strip.send(mapping['evaluator'], evaluator_text)
        end
      end

      { success: true, properties: [property_hash] }
    end

    private

    def resolve_mapping(mapping, name)
      return mapping if mapping.present?
      if name.present?
        found = ScraperMapping.find_by_name(name)
        raise ArgumentError, "Unknown scraper mapping: #{name}" unless found
        return found
      end
      # Auto-detect from URL host
      import_host = ImportHost.find_by(host: @uri.host)
      raise ArgumentError, "No mapping found for host: #{@uri.host}" unless import_host
      ScraperMapping.find_by_name(import_host.scraper_name)
    end

    def retrieve_features_array(doc, mapping, uri)
      retrieved_array = []
      if mapping['cssLocator'].present?
        css_elements = doc.css(mapping['cssLocator'])
        css_elements.each do |element|
          feature_text = get_text_from_css element, mapping
          retrieved_array.push feature_text
        end
      end

      trimmed_and_stripped_array = []
      retrieved_array.each do |string_to_clean|
        cleaned_string = clean_up_string string_to_clean, mapping
        trimmed_and_stripped_array.push cleaned_string
      end
      trimmed_and_stripped_array
    end

    def retrieve_images_array(doc, mapping, uri)
      retrieved_array = []
      if mapping['cssLocator'].present?
        css_elements = doc.css(mapping['cssLocator'])
        css_elements.each do |element|
          img_url = get_text_from_css element, mapping

          # ensure_url_is_absolute
          img_uri = URI.parse(img_url)
          unless img_uri.host
            img_uri.scheme = uri.scheme
            img_uri.host = uri.host
            unless img_uri.path.start_with? "/"
              img_uri.path = "/" + img_uri.path
            end
            if mapping['imagePathPrefix'].present?
              img_uri.path = mapping['imagePathPrefix'] + img_uri.path
            end
            img_url = img_uri.to_s
          end
          retrieved_array.push img_url
        end
      end

      if mapping['xpath'].present?
        css_elements = doc.xpath(mapping['xpath'])
        css_elements.each do |element|
          retrieved_array.push element.text
        end
      end
      trimmed_and_stripped_array = []
      retrieved_array.each do |string_to_clean|
        cleaned_string = clean_up_string string_to_clean, mapping
        trimmed_and_stripped_array.push cleaned_string
      end
      trimmed_and_stripped_array
    end

    def retrieve_target_text(doc, mapping, uri)
      retrieved_text = ''
      if mapping['scriptRegEx'].present?
        regex = Regexp.new mapping['scriptRegEx']
        retrieved_text = doc.search("script").text.scan(regex)[0] || ""
      end
      if mapping['urlPathPart'].present?
        url_path_part = mapping['urlPathPart']
        retrieved_text = get_text_from_url url_path_part, uri
      end
      if mapping['cssLocator'].present?
        css_elements = doc.css(mapping['cssLocator'])
        retrieved_text = get_text_from_css css_elements, mapping
      end
      if mapping['xpath'].present?
        css_elements = doc.xpath(mapping['xpath'])
        retrieved_text = get_text_from_css css_elements, mapping
      end
      retrieved_text = clean_up_string retrieved_text, mapping
    end

    def clean_up_string(string_to_clean, mapping)
      unless mapping['splitTextCharacter'].nil?
        begin
          splitTextCharacter = mapping['splitTextCharacter'] || ' '
          splitTextArrayId = mapping['splitTextArrayId'].to_i
          string_to_clean = string_to_clean.split(splitTextCharacter)[splitTextArrayId]
          string_to_clean = string_to_clean || ''
        rescue StandardError => e
          Rails.logger.error "PropertyWebScraper: clean_up_string failed: #{e.message}"
        end
      end
      if mapping['stripString'].present?
        string_to_clean = string_to_clean.sub(mapping['stripString'], "")
      end
      string_to_clean
    end

    def get_text_from_url(url_path_part, uri)
      url_retrieved_text = uri.path
      if url_path_part.to_i > 0
        url_retrieved_text = uri.path.split("/")[url_path_part.to_i]
      end
      url_retrieved_text
    end

    def get_text_from_css(css_elements, mapping)
      css_retrieved_text = css_elements.text
      if mapping['cssAttr'] && css_elements.attr(mapping['cssAttr'])
        css_retrieved_text = css_elements.attr(mapping['cssAttr']).to_s
      elsif mapping['xmlAttr'] && css_elements.attr(mapping['xmlAttr'])
        css_retrieved_text = css_elements.attr(mapping['xmlAttr'])
      end

      if mapping['cssCountId'].present?
        begin
          css_count_id = mapping['cssCountId'].to_i
          css_retrieved_text = css_elements[css_count_id].text || ''
        rescue StandardError => e
          Rails.logger.error "PropertyWebScraper: get_text_from_css failed: #{e.message}"
        end
      end
      css_retrieved_text
    end
  end
end
